### ISSUE TITLE
Support for arbitrary types in mappings (given a converter?)

### ISSUE THREAD

**Issue by airhorns** [imdrasil/jennifer.cr#323]
I'm trying to map a PG enum to a Crystal enum so that I can have type safety when working with the column in question. I figured I'd just write my own converter that wraps the enum-to-string conversion Jennifer already has built in to take it one step further and map the string onto the enum like so:

```crystal
class CrystalEnumConverter(T)
  def from_db(pull, nillable)
    T.parse(Jennifer::Model::EnumConverter.from_db(pull, nillable))
  end

  def self.to_db(value : T) : String
    value.to_s
  end

  def self.to_db(value : Nil)
    value
  end

  def self.from_hash(hash : Hash, column)
    T.parse(Jennifer::Model::EnumConverter.from_hash(hash, column))
  end
end
```

Then, in my model, I specified my custom type and the converter in the model mapping:

```crystal
class Product < Jennifer::Model::Base
  with_timestamps

  mapping(
    id: Primary64,
    name: String,
    type: {type: ProductType, converter: CrystalEnumConverter(ProductType).new},
  )

  belongs_to :shop, Shop
end
```

Sadly, this doesn't work because the `DBAny` type union is static. I get a compile error like so:

```
In lib/jennifer/src/jennifer/relation/belongs_to.cr:81:39

 81 | collection.each { |e| array << e.attribute(_foreign) }
                                  ^-
Error: no overload matches 'Array(Array(Char) | Array(Float32) | Array(Float64) | Array(Int16) | Array(Int32) | Array(Int64) | Array(String) | Bool | Char | Float32 | Float64 | Int16 | Int32 | Int64 | Int8 | JSON::Any | PG::Geo::Box | PG::Geo::Circle | PG::Geo::Line | PG::Geo::LineSegment | PG::Geo::Path | PG::Geo::Point | PG::Geo::Polygon | PG::Numeric | Slice(UInt8) | String | Time | Time::Span | UInt32 | Nil)#<<' with type (ProductType | Int32 | Int64 | JSON::Any | String | Time | Nil)
```

Note the `ProductType` type in the union trying to be pushed.

That `collection` variable is an `Array(DBAny)` (https://github.com/imdrasil/jennifer.cr/blob/master/src/jennifer/relation/belongs_to.cr#L80) , and `DBAny` is hardcoded inside Jennifer. 

Is this use case supported? I can think of lots of other use cases for mapping types from the database to something more useful (or type safe!) in the application code, and it'd be ideal for end users to be able to do that, but I get that it's likely difficult to support internally. If it isn't or can't be supported, I can add that to the docs too.

**Comment by imdrasil** [comment]
Technically a model attribute should be one of `Jennifer::DBAny` type therefor we can't use native enums. But maybe is we implement a bit more granular attribute accessing mechanism we will be able to remove this restriction. Will work on this on weekend :+1:

**Comment by imdrasil** [comment]
seams I'll manage to add support of arbitrary types in mapping - continue working on this

**Comment by imdrasil** [comment]
the changes are e bit more fundamental that I expected so this will take some more time

**Comment by airhorns** [comment]
Absolutely no rush or pressure, thanks for putting something as high quality as Jennifer together in the first place!

**Comment by imdrasil** [comment]
This is resolved by #352
